RFC: KScan: Add new GPIO matrix driver.

petejohanson opened the pull request:
* Timer based polling of GPIO matrix for key states.
* DTS bindings to specify column and row GPIO pins/flags.
* DTS property for controlling scanning direction depending
  on per key diode direction for anti-ghosting.

I'm exploring keyboard firmwares built on Zephyr, and one of the major items I've found lacking that's used by most mechanical keyboards is a generic GPIO matrix for detecting keypresses while minimizing the # of used GPIO pins.

The existing `kscan` driver seems *perfectly* suited to this, and so I've created a preliminary working GPIO matrix based kscan driver that is tested working with my hacked together breadboard "keyboard" and an Nucleo STM32WB dev kit.

Opening an initial PR to get some thoughts and feedback. You can see the *very* early firmware work over at https://gitlab.com/zmkproject/zmk/ and you can see an example DTS file that has the KScan settings at https://gitlab.com/zmkproject/zmk/-/blob/master/boards/shields/petejohanson_handwire/petejohanson_handwire.overlay#L39

Something that there is interest in merging upstream here?

Thanks in advance.

petejohanson committed:
KScan: Add new GPIO matrix driver.

* Timer based polling of GPIO matrix for key states.
* DTS bindings to specify column and row GPIO pins/flags.
* DTS property for controlling scanning direction depending
  on per key diode direction for anti-ghosting.

Signed-off-by: Pete Johanson <peter@peterjohanson.com>

petejohanson commented on drivers/kscan/kscan_gpio.c:
```diff
@@ -0,0 +1,320 @@
+	gpio_flags_t row_gpio_dir = GPIO_INPUT;
+#endif
+
+#if DT_PROP_HAS_IDX(MATRIX_NODE_ID, col_gpios, 0)
```
This is the one piece I am most disliking about this implementation. Because the number of entries in the `col-gpios` and `row-gpios` is dynamic, we need to loop over the to initialize/load things, but there's no easy *macro* to do so.

#24536 is an open issue about this, and having a `DT_GPIO_FOREACH` would let me write this safely with something like:

```
DT_GPIOS_FOREACH(MATRIX_NODE_ID, col_gpios, INIT_COL_SCAN_ITEM)
```

and have it "Just Work"

jhn-nordic commented:
A couple of comments based on experience with wireless keyboards: 
1. A robust way of switching between gpio interrupt-driven an polled operation is something you will need at least when making battery powered keyboards.  
2. Most consumer keyboard matrixes does not have diodes to prevent ghosting -> Some sort of ghosting detection in fw would be useful.

petejohanson commented:
> A couple of comments based on experience with wireless keyboards:
> 
> 1. A robust way of switching between gpio interrupt-driven an polled operation is something you will need at least when making battery powered keyboards.
> 2. Most consumer keyboard matrixes does not have diodes to prevent ghosting -> Some sort of ghosting detection in fw would be useful.

@jhn-nordic Thanks for your thoughts.

1. Yeah, I had a version of this that was GPIO interrupt driven, and hit some snags trying to implement it in a way that didn't cause "interrupt loops" when doing the actual scanning. It's *definitely* on my list to figure out, but wanted some early feedback on validation of the general approach. Helpful to hear that I was thinking the right way about it, even if I had problems implementing it.
1. Yeah, adding ghosting detection shouldn't be hard. My first target is definitely the more niche "mechanical keyboard audience", where most of those PCBs have ghost prevention diodes.

jfischer-phytec-iot commented on drivers/kscan/kscan_gpio.c:
```diff
@@ -0,0 +1,320 @@
+	kscan_gpio_read(data->dev);
+}
+
+#define INIT_ROW_SCAN_ITEM(idx) \
```
Except that it does not follow coding style and is not beautiful, why can you not write it as a function?

petejohanson replied:
@jfischer-phytec-iot Maybe I'm completely missing something, but see my comments on #24536 about needing to loop from 0 to max rows/columns, and generate entries in an array. In doing so, you need to get the pin/flags from the defines, so you can't loop at *runtime* to fetch it, you need to "unroll the loop" at *compile* time, AFAICT.

If I'm missing something, by all means I'd love to understand how to achieve pulling a dynamic length phandles-array out, with full details, to configure/use.

petejohanson replied:
@jfischer-phytec-iot You can see a branch I made of this work that adds a macro that eliminates the need for this ugly hack in my comment here: https://github.com/zephyrproject-rtos/zephyr/issues/24536#issuecomment-629571341

rapgenic commented:
This would be very useful! 

However it might be great to support having multiple instances in the device tree (in my application I have 3 keyboard matrices), if it is not too complicated of course.
 
Also if this is possible, would it be more reasonable to have `KSCAN_GPIO_POLL_PERIOD` as a device tree property instead of a config one?

petejohanson commented:
@rapgenic 

> However it might be great to support having multiple instances in the device tree (in my application I have 3 keyboard matrices), if it is not too complicated of course.

It shouldn't be hard to tweak this to be truly multi-instance. I can take a look at doing that.
 
> Also if this is possible, would it be more reasonable to have `KSCAN_GPIO_POLL_PERIOD` as a device tree property instead of a config one?

Hmm... I modeled this after one of the *other* KSCAN drivers in `master`. If you do it in the DT, then it's less obvious how applications can change the config... they need to override it with a DT overlay, versus just updating the Kconfig setting.

To a certain degree, that config value only exists because my first attempt at GPIO interrupt based tirggers failed. With that working that could be removed. Do you really need to very that period, or would the interrupt based scanning suffice?

jhn-nordic commented:
Depending on how much the buttons bounce it can be beneficial to be able to configure the poll period.

petejohanson commented:
@jhn-nordic So, I've pushed an update that switched to interrupt based scanning. I finally figure out what I had missed at my previous attempt. This disables the interrupts temporarily while doing the scanning to we don't trigger the ISR *again* when toggling the outputs.

I didn't bother *keeping* the poll based version, since I'm not seeing the benefit of this over an interrupt based one. Is there some reason you might want to fall back to polling? Lack of available interrupts on some platforms?

I've also tried to clean up a bunch of patch warnings on whitespace, long lines, etc.

The one remaining *wart* from a code perspective is still the row/column extraction from the DT defines, which still is "manually unrolled" for now.

In my mind, at a minimum, this still needs debouncing support (I updated this to include the `GPIO_INT_DEBOUNCE`, but not all platforms support this, so probably need to handle debouncing specifically here), but some other potential outstanding issues:

* Ghost detection as a configurable/DT option.
* Multi-instance support (this shouldn't be hard, but I'd rather do that once the core is settled/working)
* Fix for the manually "loop unrolling" pending a resolution/fix for #24536.
* Fallback polling mode, if really needed.

petejohanson commented:
Ok, added some basic debouncing with `k_delayed_work` in the ISR for the matrix. Added a new binding property for the debounce milliseconds, so it can eventually be set per-matrix, once we have multiple instances.

franciscomunoz commented on drivers/kscan/kscan_gpio.c:
```diff
@@ -0,0 +1,495 @@
+		gpio_pin_set(col, col_cfg->pin, 0);
+	}
+
+	// Set all our outputs as active again, then re-enable interrupts,
```
Change all your comments to /* */

petejohanson replied:
@fractalclone Done. (And rebased on latest `master` while at it to be safe).

franciscomunoz commented on dts/bindings/kscan/gpio-kscan.yaml:
```diff
@@ -0,0 +1,25 @@
+        required: true
+    debounce-period:
+        type: int
+        default: 5
```
It would be better if you add a Kconfig entry for your debounce period in your Kconfig.gpio. You could change tune it for other keyboard brands.  See the way it was done in Kconfig.xec and ini my opinion this parameter shouldn't be part of the dts/bindings as this is more related to keyboards you can swap and test.  There are keyboard that do not have diodes as these feature makes them more expensive. Therefore, I also believe Diode-direction should be part of your kconfig, in fact you could add another flag to specify where your keyboard has diodes.  This is just.a suggestion.

Delsian commented:
@petejohanson Thanks for this module, I have plans to use it in my project, but after cherry-picking I found some minor issues with merging to current master.
Please, can you rebase your branch to actual master?

petejohanson commented:
> @petejohanson Thanks for this module, I have plans to use it in my project, but after cherry-picking I found some minor issues with merging to current master.
> Please, can you rebase your branch to actual master?

I will work on that when I have a moment. I have been working on this more "out of tree", since I created this PR. You can find the more active version here https://github.com/zmkfirmware/zmk/blob/main/app/drivers/zephyr/kscan_gpio_matrix.c but note we are targeting 2.3.0 there, so it will have issues w/ C99 type changes, etc.

Delsian commented on drivers/kscan/kscan_gpio.c:
```diff
@@ -0,0 +1,495 @@
+#else
+	struct k_delayed_work work;
+#endif
+	bool matrix_state[MATRIX_ROWS][MATRIX_COLS];
```
Before the first interrupt matrix_state contains uninitialized values. As a result, on the very first keypress we get long sequence of callbacks for every key.
kscan_gpio_init() should call dummy kscan_gpio_read() with empty callback function.

Delsian commented:
@petejohanson Your code (after minor changes) works fine in my project. But it required to keep separate Zephyr branch with your code. If you agree, I can fix all requested changes and create a separate pull request to merge your sources into master.
